src/Groups/matrices/iso_nf_fq.jl

```diff
@@ -0,0 +1,172 @@
+# Detinko, Flannery, O'Brien "Recognizing finite matrix groups over infinite
+# fields", Section 4.2
+function isomorphic_group_over_finite_field(matrices::Vector{T}) where T <: Union{ fmpq_mat, AbstractAlgebra.MatElem{nf_elem} }
+   @assert !isempty(matrices)
+
+   # One should probably check whether all matrices are n by n (and invertible
+   # and such ...)
+
+   K = base_ring(matrices[1])
+   n = nrows(matrices[1])
+
+   Fq, matrices_Fq = good_reduction(matrices, 2)
+
+   G = MatrixGroup(n, Fq, matrices_Fq)
+   N = GAP.Globals.Order(G.X)
+   if maximal_order_of_finite_linear_group_qq(degree(K)*n) < N
+      error("Group is not finite")
+   end
+
+   G_to_fin_pres = GAP.Globals.IsomorphismFpGroupByGenerators(G.X, GAP.julia_to_gap([ g.X for g in gens(G) ]))
+   F = GAP.Globals.Range(G_to_fin_pres)
+   rels = GAP.Globals.RelatorsOfFpGroup(F)
+
+   inv_matrices = Vector{T}(undef, length(matrices))
+   inv_computed = falses(length(matrices))
+   for i = 1:length(rels)
+      rel = map(Int, GAP.Globals.LetterRepAssocWord(rels[i]))
+      M = identity_matrix(K, n)
+      for k in rel
+         if k < 0
+            k = -k
+            if !inv_computed[k]
+               inv_matrices[k] = inv(matrices[k])
+            end
+            M *= inv_matrices[k]
+         else
+            M *= matrices[k]
+         end
+      end
+      if !isone(M)
+         error("Group is not finite")
+      end
+   end
+   return G
+end
+
+function good_reduction(matrices::Vector{T}, start::Int = 2) where T <: Union{ fmpq_mat, AbstractAlgebra.MatElem{nf_elem} }
+   p = next_prime(start)
+   b, Fq, matrices_Fq = test_modulus(matrices, p)
+   while !b
+      p = next_prime(p)
+      b, Fq, matrices_Fq = test_modulus(matrices, p)
+   end
+   return Fq, matrices_Fq
+end
+
+# Detinko, Flannery, O'Brien "Recognizing finite matrix  groups over infinite
+# fields", Section 3.1 claims that any prime != 2 not dividing any denominator
+# of the matrices and their inverses (!) works.
+function test_modulus(matrices::Vector{fmpq_mat}, p::Int)
+   Fp = GF(p)
+   matrices_Fp = Vector{AbstractAlgebra.MatElem{elem_type(Fp)}}(undef, length(matrices))
+   if p == 2
+      return false, Fp, matrices_Fp
+   end
+
+   for M in matrices
+      for i = 1:nrows(M)
+         for j = 1:ncols(M)
+            if iszero(M[i, j])
+               continue
+            end
+
+            if mod(denominator(M[i, j]), p) == 0
+               return false, Fp, matrices_Fp
+            end
+         end
+      end
+   end
+   # I don't want to invert everything in char 0, so I just check whether the
+   # matrices are still invertible mod p.
```

Comment:
I wonder how expensive these rank computations get, though?

I guess it depends on how many primes we have to look at... If it is a lot, it might be better to invert the matrices once, at the start (and then pass the inverted matrices to this function)?

Alternatively, one can compute the determinants of the matrices, and reject primes dividing the numerator or denominator of those. (And *then* it might be beneficial to perform this test first, before iterating over all those matrix entries?)

But all of this is just musings for future tweaking, for now don't feel obliged to act on it :-)

Reply:
I'm not sure what's best here either. We also need to invert (some of) the matrices when we evaluate the relations of the presentation later.